File: src/cpp/decoder.cpp
#include "decoder.hpp"

#include "Barcode.h"
#include "BarcodeFormat.h"
#include "ImageView.h"
#include "ReadBarcode.h"
#include "ReaderOptions.h"

#include <opencv2/objdetect.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/opencv.hpp>

namespace qrscanner {


cv::Mat croplocate_cv2(const cv::Mat& image)
{
    if (image.empty())
        return image;

    cv::QRCodeDetector detector;

    // Немного помогает на сложных QR
    detector.setUseAlignmentMarkers(true);

    std::vector<cv::Point> points;
    bool found = detector.detect(image, points);

    // Не нашли QR -> возвращаем исходник
    if (!found || points.size() != 4)
        return image;

    // Bounding rect вокруг найденного quad
    cv::Rect bbox = cv::boundingRect(points);

    // Добавляем padding
    int pad_x = static_cast<int>(bbox.width * 0.15);
    int pad_y = static_cast<int>(bbox.height * 0.15);

    bbox.x -= pad_x;
    bbox.y -= pad_y;
    bbox.width += pad_x * 2;
    bbox.height += pad_y * 2;

    // Клип к границам изображения
    bbox &= cv::Rect(0, 0, image.cols, image.rows);

    if (bbox.width <= 0 || bbox.height <= 0)
        return image;

    return image(bbox).clone();
}

static cv::Mat continuous_for_zxing(const cv::Mat& img, ZXing::ImageFormat& format)
{
    cv::Mat view = img;
    if (img.channels() == 1) {
        format = ZXing::ImageFormat::Lum;
    } else if (img.channels() == 3) {
        format = ZXing::ImageFormat::BGR;
    } else if (img.channels() == 4) {
        format = ZXing::ImageFormat::BGRA;
    } else {
        cv::cvtColor(img, view, cv::COLOR_BGR2GRAY);
        format = ZXing::ImageFormat::Lum;
    }
    if (!view.isContinuous())
        view = view.clone();
    return view;
}

std::optional<std::string> decode_cv2(const cv::Mat& img)
{
    cv::Mat gray;
    if (img.channels() == 1)
        gray = img;
    else
        cv::cvtColor(img, gray, cv::COLOR_BGR2GRAY);

    cv::QRCodeDetector detector;
    std::string data = detector.detectAndDecode(gray);
    if (!data.empty())
        return data;
    return std::nullopt;
}

std::optional<std::string> decode_zxingcpp(const cv::Mat& img)
{
    ZXing::ImageFormat format = ZXing::ImageFormat::None;
    cv::Mat buffer = continuous_for_zxing(img, format);
    ZXing::ImageView view(buffer.data,
                          static_cast<int>(buffer.total() * buffer.elemSize()),
                          buffer.cols,
                          buffer.rows,
                          format,
                          static_cast<int>(buffer.step),
                          static_cast<int>(buffer.elemSize()));

    ZXing::ReaderOptions options;
    options.setFormats(ZXing::BarcodeFormat::QRCode)
           .setTryHarder(true)
           .setTryRotate(true)
           .setTryInvert(true)
           .setTryDownscale(true)
           .setTextMode(ZXing::TextMode::HRI);

    auto result = ZXing::ReadBarcode(view, options);
    if (result.isValid())
        return result.text();
    return std::nullopt;
}

static std::optional<std::string> try_decode(const cv::Mat& img)
{
    auto result = decode_zxingcpp(img);
    if (result && !result->empty())
        return result;
    return std::nullopt;
}

static std::optional<std::string> try_gray_variants(const cv::Mat& gray)
{
    // if (auto r = try_decode(gray); r)
    //     return r;

    cv::Mat binary;
    cv::threshold(gray, binary, 0, 255, cv::THRESH_BINARY | cv::THRESH_OTSU);
    if (auto r = try_decode(binary); r)
        return r;

    cv::Mat blurred;
    cv::GaussianBlur(gray, blurred, cv::Size(5, 5), 0.0);
    if (auto r = try_decode(blurred); r)
        return r;

    cv::GaussianBlur(gray, blurred, cv::Size(7, 7), 0.0);
    if (auto r = try_decode(blurred); r)
        return r;

    cv::Mat sharpened;
    static const cv::Mat sharpen_kernel = (cv::Mat_<float>(3, 3) <<
        -1.f, -1.f, -1.f,
        -1.f,  9.f, -1.f,
        -1.f, -1.f, -1.f
    );

    cv::filter2D(gray, sharpened, -1, sharpen_kernel);
    if (auto r = try_decode(sharpened); r)
        return r;

    cv::threshold(sharpened, binary, 0, 255, cv::THRESH_BINARY | cv::THRESH_OTSU);
    if (auto r = try_decode(binary); r)
        return r;

    cv::GaussianBlur(sharpened, blurred, cv::Size(5, 5), 0.0);
    if (auto r = try_decode(blurred); r)
        return r;

    cv::GaussianBlur(sharpened, blurred, cv::Size(7, 7), 0.0);
    if (auto r = try_decode(blurred); r)
        return r;

    return std::nullopt;
}

std::optional<std::string> decode_qreader(const cv::Mat& img)
{
    // return std::nullopt;
    if (img.empty() || img.depth() != CV_8U)
        return std::nullopt;

    // if (auto r = try_decode(img); r)
    //     return r;

    // cv::Mat cropped = croplocate_cv2(img);
    // cv::imwrite("debug/cropped.png", cropped);

    cv::Mat inverted;
    cv::bitwise_not(img, inverted);
    if (auto r = try_decode(inverted); r)
        return r;

    // cv::Mat gray;
    // if (img.channels() == 1) {
    //     gray = img;
    // } else if (img.channels() == 3) {
    //     cv::cvtColor(img, gray, cv::COLOR_BGR2GRAY);
    // } else if (img.channels() == 4) {
    //     cv::cvtColor(img, gray, cv::COLOR_BGRA2GRAY);
    // } else {
    //     cv::cvtColor(img, gray, cv::COLOR_BGR2GRAY);
    // }

    return try_gray_variants(img);
}

std::optional<std::string> decode(const cv::Mat& img)
{
    auto result = decode_zxingcpp(img);
    if (result)
        return result;
    return std::nullopt;
}

} // namespace qrscanner

File: src/python/decoder.py
from __future__ import annotations
import cv2
from PIL import Image
import numpy as np
from pyzbar.pyzbar import decode as pyzbar_decode
from random import random
from time import time
import zxingcpp
from tools import *


from typing import Callable, Any, Iterable
import cv2
import numpy as np

_SHARPEN_KERNEL = np.array(
    [
        [-1.0, -1.0, -1.0],
        [-1.0,  9.0, -1.0],
        [-1.0, -1.0, -1.0],
    ],
    dtype=np.float32,
)

def qreader_style_global_decode(
    image: np.ndarray,
    decode: Callable[[np.ndarray], Any],
) -> Any | None:
    """
    Повторяет глобальные preprocessing-этапы из QReader и передает каждый
    вариант в вашу функцию decode(image).

    Ожидается RGB uint8 изображение HxWxC или HxW.
    decode(image) -> результат декодирования, либо None/False при неудаче.
    """

    if image.dtype != np.uint8:
        raise ValueError(f"Expected uint8 image, got {image.dtype}")

    # Handle different channel configurations
    if image.ndim == 3:
        if image.shape[2] == 1:
            # Single channel with explicit dimension
            image = image[:, :, 0]
        elif image.shape[2] == 4:
            # RGBA/BGRA -> convert to RGB by removing alpha channel
            image = image[:, :, :3]
        elif image.shape[2] != 3:
            raise ValueError(f"Expected 1, 3, or 4 channels, got {image.shape[2]} channels")

    def _try(img: np.ndarray) -> Any | None:
        result = decode(img)
        return result if result is not None and result is not False else None

    def _threshold_and_blur_decodings(gray_or_color: np.ndarray) -> Any | None:
        # 1) original
        result = _try(gray_or_color)
        if result is not None:
            return result

        # 2) Otsu binarization only for grayscale, exactly like QReader
        if gray_or_color.ndim == 2:
            _, binary_image = cv2.threshold(
                gray_or_color,
                thresh=0,
                maxval=255,
                type=cv2.THRESH_BINARY + cv2.THRESH_OTSU,
            )
            result = _try(binary_image)
            if result is not None:
                return result

        # 3) Gaussian blur variants
        for kernel_size in ((5, 5), (7, 7)):
            blurred = cv2.GaussianBlur(gray_or_color, kernel_size, sigmaX=0)
            result = _try(blurred)
            if result is not None:
                return result

        return None

    variants: list[np.ndarray] = [image]

    # Inverted image
    variants.append(np.array(255, dtype=np.uint8) - image)

    # Grayscale version, same channel assumption as QReader: RGB -> gray
    if image.ndim == 3:
        gray = cv2.cvtColor(image, cv2.COLOR_RGB2GRAY)
    else:
        gray = image
    variants.append(gray)

    # Sharpened grayscale
    if image.ndim == 3:
        sharpened = cv2.filter2D(src=image, ddepth=-1, kernel=_SHARPEN_KERNEL)
        sharpened_gray = cv2.cvtColor(sharpened, cv2.COLOR_RGB2GRAY)
    else:
        sharpened_gray = cv2.filter2D(src=image, ddepth=-1, kernel=_SHARPEN_KERNEL)
    sharpened_gray = np.clip(sharpened_gray, 0, 255).astype(np.uint8)
    variants.append(sharpened_gray)

    # Apply the same progression as QReader: original -> inverted -> grayscale+filters -> sharpened+filters
    for idx, variant in enumerate(variants):
        if idx < 2:
            result = _try(variant)
            if result is not None:
                return result
        elif idx == 2:
            result = _threshold_and_blur_decodings(variant)
            if result is not None:
                return result
        else:
            result = _threshold_and_blur_decodings(variant)
            if result is not None:
                return result

    return None

def decode_pyzbar(img): #0.0012s
    results = pyzbar_decode(img)
    return results[0].data.decode('utf-8') if results else None


def decode_cv2(img): #0.007s
    img = np.array(img)
    
    if len(img.shape) >= 3:
        img = cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)
    
    detector = cv2.QRCodeDetector()  
    data, points, _ = detector.detectAndDecode(img)
    
    if not data and points is not None:
        data, _ = detector.decodeCurved(img, points)
    
    return data or None

from qreader import QReader
qreader = QReader(model_size="n")
def decode_qreader(img): #0.46
    img = np.array(img)
    try:
        return qreader.detect_and_decode(image=img)[0]
    except:
        return None

def decode_zxingcpp(img):
    try:
        result = zxingcpp.read_barcode(img).text
        return result
    except:
        return None

# log = [0, 0, 0]
def decode(img):
    # print(log)
    # img = cv2.cvtColor(img_bgr, cv2.COLOR_BGR2RGB)
    # img = Image.fromarray(img)
    result = decode_zxingcpp(img)
    if result is not None:
        # log[0] += 1
        return result
    
    return None
    
    result = decode_pyzbar(img)
    if result is not None:
        # log[1] += 1
        return result
    
    if random() < 0.02:
        result = decode_qreader(img)
        if result is not None:
            # log[2] += 1
            return result


if __name__ == "__main__":

    
    img = Image.open("debug/corrected_full_tps.png")
    
    # weight, height = img.size
    # scale = min(1000 / weight, 1000 / height)
    # img = img.resize((int(weight * scale), int(height * scale)))
    

    # start = time()
    # for i in range(100):
    #     decode_cv2(img)
    # print((time() - start) / 100)
    
    # show_points(img, [])
    print(decode_pyzbar(img))
    
    result = qreader_style_global_decode(np.array(img), decode_pyzbar)
    print(result)
    
    
    result = decode_zxingcpp(img)
    print(result)
    result = decode_qreader(img)
    print(result)